### DISCUSSION TITLE
关于generator实验函数体返回值问题

### DISCUSSION THREAD

**Discussion by SundayDream00** [arcsysu/SYsU-lang#37] [Q&A] [answered]
助教好，对于模板中函数IR生成部分的返回语句有一些问题想问，模板代码是这样的：
```
if (auto RetVal = llvm::ConstantInt::get(
          TheContext, /* i32 3(decimal) */ llvm::APInt(32, "3", 10))) {
    // Finish off the function.
    Builder.CreateRet(RetVal);
```
问题如下：
1. 代码中的数字“3”是否因为范例测试程序原语句为`return 3;`，则此处直接指定获取`int`值为3？
2. 若1中的理解是正确的，是否意思是`CreateRet`语句需要的参数就是`return`语句返回的表达式的值呢？如：`return a;`，则`RetVal`就取 a 的值？
3. 如果想获取普适性的函数返回值，参考了万花筒和 mukulrathi 教程，是需要得到最后一个表达式的`Value`，本想设置`buildReturnStmt`进行处理，获取子节点值，但是教程程序中好像并没有`return`相关的`expr`处理，是不是我的理解有问题呢？这个返回值应该怎样获取呢？

**Comment by Yorkking** [comment, marked as answer]
1，2 理解正确。问题 3 的意思我没有理解😐，你想问表达式的 `Value` 如何生成吗？

**Reply by SundayDream00** [reply]
抱歉写的有点乱，如果2理解正确的话其实是想问那`RetVal`应该怎样获取呢？因为`return`语句可能返回的是一个表达式，或者一个`id`等等

**Reply by Yorkking** [reply]
我举例吧：
-  `return 1;` 这样的语句，那么根据 `json` 语法树信息 `{kind:IntegerLiteral, type:int}` , `RetVal` 应该使用 `llvm::ConstantInt::get(
          TheContext, lvm::APInt(32, somval , 10)); ` 生成；
- `return a;` 这样的语句，那么根据 `json` 语法树信息 `{kind: DeclRefExpr, referencedDecl:{"kind": "VarDecl","name": "a"}}`, 需要根据 `kind: DeclRefExpr`  从而得出需要从 `a` 里面取出数来返回，即通过 Builder.CreateLoad() 指令来取出对应的 `a` 的数据，而 `a` 对应的 `llvm::Value*` 怎么找呢？如果是局部变量，你可以构建一个符号表来获得，如果是全局变量，可以使用全局变量的 API.
- ...
 总之，return Exp; 你需要做的就是对 Exp 的语法树，根据不同的 `kind` 编写不同的生成 IR 的代码。

**Reply by SundayDream00** [reply]
哇好详细！明白意思了，非常感谢！

**Reply by Aurora-1412** [reply]
> 我举例吧：
> 
> * `return 1;` 这样的语句，那么根据 `json` 语法树信息 `{kind:IntegerLiteral, type:int}` , `RetVal` 应该使用 `llvm::ConstantInt::get( TheContext, lvm::APInt(32, somval , 10)); ` 生成；
> * `return a;` 这样的语句，那么根据 `json` 语法树信息 `{kind: DeclRefExpr, referencedDecl:{"kind": "VarDecl","name": "a"}}`, 需要根据 `kind: DeclRefExpr`  从而得出需要从 `a` 里面取出数来返回，即通过 Builder.CreateLoad() 指令来取出对应的 `a` 的数据，而 `a` 对应的 `llvm::Value*` 怎么找呢？如果是局部变量，你可以构建一个符号表来获得，如果是全局变量，可以使用全局变量的 API.
> * ...
>   总之，return Exp; 你需要做的就是对 Exp 的语法树，根据不同的 `kind` 编写不同的生成 IR 的代码。

我是这样取a对应的数据的：
```
auto symbolTable = BB->getValueSymbolTable(); //获得本块的符号表
auto valPointer = symbolTable->lookup(varName); //查找a对应的指针
auto v = Builder.CreateLoad(valPointer); //取出a的值
```
然后我就不知道接下来如何创建return语句了（不知道那个“somval”该填什么东西），CreateLoad的返回值（LoadInst*类型）在这里怎么用呢

**Reply by Yorkking** [reply]
直接 Builder.CreateRet(v) 就行了。

**Reply by Aurora-1412** [reply]
> 直接 Builder.CreateRet(v) 就行了。

非常感谢！

**Reply by SundayDream00** [reply]
> 直接 Builder.CreateRet(v) 就行了。

在`return exp;`部分我本想采用针对不同类型的`exp`，如数字、`id`、计算表达式等，来单独构建不同的`llvm::Value`函数进行处理，然后在`Funcbuild`部分直接使用返回值进行`CreateRet(Val)`，但是有个问题是这样的话是否就不能使用`BB->getValueSymbolTable()`获取符号表了呢？因为该块是在`func`部分被定义的，`BB`不是全局变量，请问那应该怎么处理呢？
另外对于`Int`节点，如果想获取该`Object`的`Value`值转换为`APInt`应该采用什么库函数呢？我原本构想的实现如下：
```
llvm::Value *IntegerExpr (const llvm::json::Object *O) {
  auto value = O->get("value")->getAsInteger();
  return llvm::ConstantInt::get(TheContext,llvm::APInt(value));
}
```
但是这样会报错`error: no matching conversion for functional-style cast from 'llvm::Optional<long>' to 'llvm::APInt'
  return llvm::ConstantInt::get(TheContext,llvm::APInt(value));`

**Reply by Yorkking** [reply]
> Q: 但是有个问题是这样的话是否就不能使用BB->getValueSymbolTable()获取符号表了呢？因为该块是在func部分被定义的，BB不是全局变量，请问那应该怎么处理呢？

A: 那就设成全局变量，或者把这个变量传入到你单独构建的函数里面；

> Q: 另外对于Int节点，如果想获取该Object的Value值转换为APInt应该采用什么库函数呢？

A: 从你报错来看，我猜应该这样：return llvm::ConstantInt::get(TheContext,llvm::APInt(*value));

**Reply by SundayDream00** [reply]
> > Q: 但是有个问题是这样的话是否就不能使用BB->getValueSymbolTable()获取符号表了呢？因为该块是在func部分被定义的，BB不是全局变量，请问那应该怎么处理呢？
> 
> A: 那就设成全局变量，或者把这个变量传入到你单独构建的函数里面；
> 
> > Q: 另外对于Int节点，如果想获取该Object的Value值转换为APInt应该采用什么库函数呢？
> 
> A: 从你报错来看，我猜应该这样：return llvm::ConstantInt::get(TheContext,llvm::APInt(*value));

收到，这个`Value`转换部分通过另一种方式解决了。

但是关于处理`identifier`的独立函数中这部分程序：
> auto symbolTable = BB->getValueSymbolTable(); 
> auto valPointer = symbolTable->lookup(varName);
> auto v = Builder.CreateLoad(valPointer);

我将`Basic block *BB`传入了该函数，会出现如下报错：
```
error: member access into incomplete type 'llvm::ValueSymbolTable'
  auto valPointer = symbolTable->lookup(id); 
/usr/lib/llvm-11/include/llvm/IR/SymbolTableListTraits.h:41:7: note: forward declaration of 'llvm::ValueSymbolTable'
class ValueSymbolTable;
```
已经在头文件中包含了`BasicBlock.h`和`SymbolTableListTraits.h`，请问是遗漏了什么定义或文件，或者什么其他问题导致该报错呢？

**Reply by Yorkking** [reply]
这样：`#include <llvm/IR/ValueSymbolTable.h>` 试试？

**Reply by SundayDream00** [reply]
> 这样：`#include <llvm/IR/ValueSymbolTable.h>` 试试？

原来是这个文件，感谢!
